Bugfixes in webdav.py

cfenell opened the pull request:
WebDAV module corrected and tested on Debian and Ubuntu.

Main change: Default.stat() method should report filesize correctly.

cserf committed:
BB8 rewrite.

It introduces the new BB8 daemon that replaces the scripts for nuclei and t2 background rebalancing

Include tests to validate the new daemon behaviour

Rewrite test_list_by_length that is failing

joeldierkes committed:
Core: allow `move_rule` to override all options Fix #4995

Before this commit, `move_rule` could only override two attributes of the new
rule. After the commit, `move_rule` can override every field of the new rule with
the argument `override`.

cserf committed:
Enable the possibility to add several options on the rebalancing query

Commit:
Deletion: temp tables to fetch constituents for removal

Commit:
Transfers: don't consider intermediate hops for submission

When a submitter creates an intermediate hop, this hop can be
immediately picked up by another submitter who will try to submit
it as a standalone transfer. This is not a desired behavior,
intermediate hops must not be submitted on their own.

Commit:
Transfers: source replica expression on intermediate hops. Closes #5170

This shouldn't be necessary, but it's worth adding an additional
protection in case the intermediate hop is submitted alone by error.

Commit:
Transfers: try to detect concurrent multihop submissions. Closes #5028

If creation of intermediate hops fails, don't immediately consider
the request as failed. It's probable that the problem is temporary
because of another submitter working on the same partition. Just
skip the request for some time.

joeldierkes committed:
Documentation: Unify the contribution Guide #5176

The contribution guide is duplicated. It exists in `rucio/CONTRIBUTING.rst` and
in the `documentation` project. Over time, these files could diverge, which is
not desirable.

This commit makes the CONTRIBUTING file in `documentation` the main guide for
contributions and changes all references to it.

Commit:
Clients: do not set default transfer speed in download-pfn. Closes #5185

Otherwise the following warning is always printed, even when the option
is not set:
Download with --pfn doesn't support --transfer-speed-timeout

Commit:
Clients: fix exception on download --pfn. Closes #5178

dchristidis committed:
WebUI: Refine use of Datepicker widgets #4184

bari12 committed:
Merge pull request #5189 from joeldierkes/patch-5188-Job_check_actions_fails_with__fatal__Need_to_specify_how_to_reconcile_divergent_branches__

Testing: Do not pull master in add_header script #5188

yiiyama committed:
Rules: A function to transform a complex RSE expression into single-RSE expressions #4660 (#4858)

* Rules: Cutting out a function to transform a complex RSE expression into single-RSE expressions from transmogrifier #4660

* fixed import and dealt with session leaks

joeldierkes committed:
Pre commit script (#4971)

* Add a verbose cli arument to the `add_header` script

The `add_header` script prints the generated header to the console per
default. However, most of the time this is not useful, since the change itself
is not important. The fact that the header changed is.

This commit disables the output of these unnecessary informations per default
and introduces the command line argument `--verbose` to display the entire
generated header when nesseccary.

* Add pre-commit-config

Before creating a commit the changed files should be checked for thier
integrity. All linter errors should be fixed and the header should be generated.

This commit introduces the `pre-commit` config for the pre-commit git hook,
which automatically runs flake8 and the add_header script before a commit. This
prevents developers from submitting non-conform code changes. The config is used
by the python `pre-commit` program.

ivukotic committed:
adding permissions and endpoint (#5036)

* adding permissions and endpoint

* fixing return codes. cleanup

* cleanup

Co-authored-by: Ilija Vukotic <ivukotic@cern.ch>

joeldierkes committed:
Rules: Fix syntax errors

flake8 throws some syntax errors:
```
===============================
Running flake8
===============================
lib/rucio/core/rse_selector.py:252:1: W293 blank line contains whitespace
lib/rucio/core/rse_selector.py:259:1: W293 blank line contains whitespace
lib/rucio/core/rse_selector.py:264:1: W293 blank line contains whitespace
lib/rucio/core/rse_selector.py:271:1: W293 blank line contains whitespace
lib/rucio/core/rse_selector.py:275:1: W293 blank line contains whitespace
lib/rucio/core/rse_selector.py:281:1: W293 blank line contains whitespace
lib/rucio/daemons/transmogrifier/transmogrifier.py:344:1: W293 blank line contains whitespace
lib/rucio/daemons/transmogrifier/transmogrifier.py:377:1: W293 blank line contains whitespace
```

They were introduced in d04aaaea22219edf111c4a8288618173254773de

bari12 committed:
Merge pull request #5194 from joeldierkes/fix-syntax

Rules: Fix syntax errors

Commit:
Replicas: fix geoip sorting. Closes #5133

The bug was introduced in #4968: _sort_pfns_by_priority was called,
but its return value wasn't used to actually sort the replicas.

Add a test using a small geoip-city test database. To add this test,
the database download/extraction/usage code was refactored a bit,
but there weren't any particular issues with that code.

joeldierkes committed:
Documentation: Delete redundant README files #5193

Due to the fact that rucio consists of multiple components, different `README`
files have emerged. Historically, most of the information was transfered to the
documentation. As a consequence we don't need multiple `README`s anymore, since
all necessary information got transfered into one `README` file.

joeldierkes committed:
REST & API: Change the `archive` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

joeldierkes committed:
REST & API: Change the `config` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

cserf committed:
Dark reaper improvements due to DB load : Closes #5192

joeldierkes committed:
REST & API: Change the `credentials` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

joeldierkes committed:
REST & API: Change the `dirac` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

joeldierkes committed:
Testing: Fix add_header job bug #5197

The GitHub Checkout Actions checks out the wrong commit when the commit is
force-pushed. This error is known and reported
already. (https://github.com/actions/checkout/issues/299)

To fix this error, we use the sha provided by the pull request.

Commit:
Tests: add missing call to finisher in conveyor test

Otherwise the ranking is not decreased. It worked till now by
chance, because there is an implicit call to finisher earlier in
the code which may be enough if we are lucky.

Commit:
Transfers: delete fts-throttler. Closes #5203

It looks very much like dead code.

joeldierkes committed:
REST & API: Change the `export` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

bari12 committed:
Merge pull request #5183 from joeldierkes/patch-5176-Duplicated_Contributing_guide

Documentation: Unify the contribution Guide #5176

bari12 committed:
Merge pull request #5191 from joeldierkes/feature-5111-Closed_datasets_with_NULL_bytes_or_events

Core: Ensures that `__resolve_bytes_length_events_did` returns no None values #5111

bari12 committed:
Merge pull request #5199 from joeldierkes/feature-5197-Add_header_lint_job_checks_out_wrong_commit

Testing: Fix add_header job bug #5197

bari12 committed:
Merge pull request #5195 from rcarpa/patch-5133-stabilize_geoip

Replicas: fix geoip sorting. Closes #5133

bari12 committed:
Merge pull request #5187 from rcarpa/patch-5178-attribute_error_download_pfn

Clients: fix exception on download --pfn. Closes #5178

bari12 committed:
Merge pull request #5186 from rcarpa/patch-5185-download-pfn-default-timeout

Clients: do not set default transfer speed in download-pfn. Closes #5185

maany committed:
Traces: add put_schema Closes #5156 (#5159)

* traces: add put_schema #5165

* traces: update put schema pq field #5156

* traces: update put schema mappings #5165

* traces: patch headers for trace.py #5165

Commit:
Deletion: use temporary tables to list replicas to delete. Closes #4897

Due to reasons described in detail in the linked ticket, I didn't
find any way to integrate the "limit" argument into the initial
query. We are forced to "over-shoot" and get more data from the
database. Some of these retrieved rows will be skipped latter
because they don't match additional conditions required for safe
removal (no sources; no pending transfers for last replicas).

The workflow is however completely re-worked:
- We still have an unbound query which is iterated in bulks via
`yield_per`, but this iteration now depends on the limit argument.
- We use a temporary table on the database side to help with
filtering bulks rather than performing a separate query for each
request.

Commit:
Deletion: homogenize if/else in list_and_mark_unlocked_replicas. #4897

The commit changes the behavior of the function, but it seems more
correct this way.

First, it doesn't seem logical to check the obsolete condition inside
`if state != ReplicaState.UNAVAILABLE'. Moreover, this is only done for
the case when a file has more than one replica, which is even more
confusing. It looks more correct to move the check outside the if,
so that tombstone is also checked for unavailable replicas.

Second, it's true that we don't want to count the unavailable replicas
size into total_bytes. These replicas don't exist on disk, so deleting
them will not free any space. But why isn't it done for the branch in
which we analyze the last existing replica? Change the code to have
the same behavior when a single replica exists.

Third, I don't understand why we don't count the size of obsolete
replicas into total_bytes. Move total_bytes incrementation outside
`if tombstone != OBSOLETE`.

Fourth, seems wrong to not count unavailable replicas in the
"total_files". According to function definition, the "limit" argument
defines how many replicas should be returned by the function.

bari12 committed:
Merge pull request #5175 from rcarpa/patch-5170-source_replica_expression_multihops

Transfers: rework concurrent multihop handling. Closes #5170. Closes #5028

bari12 committed:
Merge pull request #5202 from cserf/patch-5192-Dark_reaper_improvements_due_to_DB_load

 Dark reaper improvements due to DB load : Closes #5192

joeldierkes committed:
REST & API: Change the `heartbeats` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

joeldierkes committed:
REST & API: Change the `identity` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

joeldierkes committed:
Documentation: Introduce the OpenApi spec generator #5209

Currently, the rest api documentation is in a bad state. The data is obsolete
and improperly displayed.

This commit introduces the OpenApi specification as backbone for the rest api
documentation. The OpenApi specification is well defined and has a lot of
features. It is commonly used and has a modularized design, which makes it easy
to swap different components.

The specification file gets generated from the python doc comments on the flask
endpoints.

Commit:
Replicas: make geoip expire delay configurable. Closes #5210

Commit:
Conveyor: fix exception in submitter. Closes #5212

TypeError: '<' not supported between instances of 'datetime.datetime' and 'NoneType'

Commit:
Core & Internals: allow external cache for auth tokens. Closes #5297

Allow setting auth tokens in memcached, but make it configurable and
default this behavior to disabled.

Also reduce the duration of token, for #5303

cserf committed:
Core : Bug in the filter_thread_work algorithm for Postgres backend : Closes #5207

bari12 committed:
Merge pull request #5213 from rcarpa/patch-5212-fix_exception_submitter

Conveyor: fix exception in submitter. Closes #5212

Commit:
Tests: get the correct request after finisher run

Otherwise the source test is done on the incorrect request.

bari12 committed:
Merge pull request #4556 from dchristidis/patch-4184-WebUI__Refine_use_of_Datepicker_widgets

WebUI: Refine use of Datepicker widgets #4184

bari12 committed:
Merge pull request #5204 from rcarpa/feature-5203-delete_fts_throttler

Transfers: delete fts-throttler. Closes #5203

ericvaandering committed:
Fix id/id_ and some PEP-8 issues (#5136)

* Refactor stomp connection utility

* Update cache consumer to
Fix id/id_ and some PEP-8 issues
and  get non-SSL connection

Redid the stomp code to look more like Kronos

* Try label

bari12 committed:
Merge pull request #5184 from joeldierkes/feature-5172-Remove_forecast_extension_and_numpy_as_a_dependency

Extensions: Remove the forecast extension

joeldierkes committed:
Clients: Add client cert and key env variables #5216

If the user authenticates via the X509 protocol, a valid certification and key
file has to be provided. Currently, the path to these files can only be set via
the config file.

This commit introduces two environment variables to set the path of the
respective files. If these variables are not provided, the ones specified in the
config will be used.

bari12 committed:
Merge pull request #5211 from rcarpa/patch-5210-configurable-geoip-expire-delay

Replicas: make geoip expire delay configurable. Closes #5210

bari12 committed:
Merge pull request #5217 from joeldierkes/feature-5216-user_environment_variables_to_override_client_cert_and_client_key_of_rucio_cfg

Clients: Add client cert and key env variables #5216

bari12 committed:
Merge pull request #5218 from rcarpa/patch-0-tests_conveyor_get_request_after_finisher

Tests: get the correct request after finisher run. Closes #5190

Commit:
Tests: split test_multihop_intermediate_replica_lifecycle. Closes #5190

In this test, submitter is executed with bulk=1. As a consequence,
the two hops of the multihop are submitted as 2 independent
transfers. This is the desired effect, as it allows to test the
recovery of the second hop. And also the fact that reaper doesn't
remove intermediate replicas.

However, if fts is slow, it's possible that the second hop is not
tried by fts immediately, but after a certain time. This results in the
second transfer to succeed while the test expects it to fail. As github
runners don't seem particularly rapid lately, this has a high chance to
happen. Especially because this is the first conveyor test and it
submits a couple of unrelated transfers (test:file[1,2,3,4] from
https://github.com/rucio/rucio/blob/1d57db0a417dc094cc06e5a9db34c5bb1b97a47b/tools/docker_activate_rses.sh#L118 )

Change the test to not relly on the fact that FTS fails the second
transfer anymore. To avoid reducing the coverage, move the reaper
test of intermediate replicas to another test.

bari12 committed:
Merge pull request #5215 from cserf/patch-5207-Core___Bug_in_the_filter_thread_work_algorithm_for_Postgres_backend

Core : Bug in the filter_thread_work algorithm for Postgres backend : Closes #5207

bari12 committed:
Merge pull request #5196 from joeldierkes/feature-5193-Delete_redundant_README_files

Documentation: Delete redundant README files #5193

bari12 committed:
Merge pull request #5221 from rcarpa/patch-5190-conveyor_test_flakiness

Tests: split test_multihop_intermediate_replica_lifecycle. Closes #5190

Commit:
Transfers: homogenize daemon infinite loop handling. Closes #5231

Each conveyor daemon had its own copy-pasted version of the same
loop: with minor adjustements which were at different level of quality
depending on when (and from which other daemon) the loop was copied.

Create a run_conveyor_daemon function which is used now by all daemons.
The common loop is based on the one previously used in submitter:
using a priority queue to avoid busy looping and for prioritizing the
activities which actually have work to do.

Rework all daemons to have a run_once function which is transmitted
as a callback. Some daemons already had such function, so it was
slightly adapted to do the work. For other daemons it's just code move.

bari12 committed:
Database: Adding query hint to list_dids; Fix #5225

bari12 committed:
Merge pull request #5229 from bari12/patch-5225-Database__Add_query_hint_to_list_dids

Database: Adding query hint to list_dids; Fix #5225

bari12 committed:
Merge pull request #5232 from rcarpa/patch-0-conveyor_daemon_loop_rework

Transfers: homogenize daemon infinite loop handling. Closes #5231

robbarnsley committed:
fixing merge conflict

robbarnsley committed:
Merge branch 'master' into feature-4949-Added_client_REST_api_route_for_listing_and_getting_historical_requests

robbarnsley committed:
Changes following Radu's comments.

robbarnsley committed:
ran tools/add_header for changed files

jamesp-epcc committed:
Bug: Fix objectstore uploads on multi VO #5235

robbarnsley committed:
removed double definition of get_request_history_by_did in core.py

robbarnsley committed:
transfers: changed config_get to config_get_bool #5219

robbarnsley committed:
adding permission restrictions to multi-vo

cserf committed:
Core : Remove hard coded value for GCS storage elements : Closes #5222

bari12 committed:
Merge pull request #5240 from robbarnsley/patch-5219-fix_allow_user_oidc_tokens_type

Transfers: changed config_get to config_get_bool #5219

Commit:
Transfers: fix metric name globus poller

prometheus doesn't support having both counters (see line 640)
and timers with the same name.

maany committed:
testing: disable test_multihop_intermediate_replica_lifecycle #5243

Commit:
Core & Internals: delete unused query_requests functions

bari12 committed:
Merge pull request #5244 from maany/feature-5243-disable_test_conveyor

Testing: disable test_conveyor #5243

Commit:
Core & Internals: get rid of FTS3Transfertool in request core

To be able to use request core in transfertool classes without
circular import problems.

It also fits more correctly the initial desire to have everything
related to transfer handling in transfer.py.

joeldierkes committed:
Testing: Print error logs in the integration test actions #5253

Commit:
Core & Internals: re-organise transfer code

The motivation for this change comes from a circular import
dependency when I tried to import transfer.py into rules.
When a rule is cancelled, it needs to cancel the Transfer in
FTS. And it seems right to put the code which cancels the
transfer in transfer.py.

This commit only does code move: some of the functions which
don't need interaction with transfertools are moved to
request.py. RSE info retrieval is also moved out. What is
clearly only needed for submitters was moved to conveyor/
common.py.

bari12 committed:
Merge pull request #5246 from rcarpa/patch-0-minor_globus_poller_fixes

Transfers: fix metric name globus poller

martynia committed:
Make Dirac server plugin Multi-VO capable (#5125)

* add a simple, default dirac scope extraction algorithm; add VO handling to dirac.py (flaskapi/v1, api and core)

* fix headers

* Move session arguments to the end in add_files() signature; move session arg to the end in list_scopes call

joeldierkes committed:
Set the itsdangerous module version prior to 2.1.0 #5258

The new `itsdangerous` package version 2.1.0 is incompatible with the Flask
version 1.1.2.

maany committed:
traces: bulk schema fixes for traces #5265

bari12 committed:
Merge pull request #5259 from joeldierkes/patch-5258-Flask_fails_after_itsdangerous_module_s_update

 Set the itsdangerous module version prior to 2.1.0 #5258

Commit:
Transfers: disentangle fts3 code from poller and receiver. Closes #4873

Until this commit, poller and receiver were filling a 'response'
dictionary and passing it back to core code. It contained:
- data needed to update the requests in the database, and this data was
modified "in place" by multiple functions in the call stack.
- temporary data used only by core function to pass a kind of "context"
through the call stack
- data only needed to generate the hermes messages and not having any
functional impact on conveyor.

As this dictionary was used in multiple places for different purposes,
many code paths filled the dictionary more or less well and in
different ways. It is very difficult to track what actually is in that
dictionary at the end of each code path. For this reason, I'm not 100%
sure I didn't introduce any regressions. My doubt is mostly about the
content of hermes messages, as the functional behavior is relatively
well tested in this part of rucio.

This commit gets rid of the dictionary. Transfertool code now returns
polymorphic objects. These objects can be queried for what fields have
to be updated in the database and which have to be sent to hermes.
I moved many transfertool specifics to these new classes.

This commit should slightly reduce the load on the database because
multiple repeated retrievals of requests by request_id from database
are now merged and done, at most, once.

bari12 committed:
Merge pull request #5245 from cserf/patch-5222-Core___Remove_hard_coded_value_for_GCS_storage_elements

Core : Remove hard coded value for GCS storage elements : Closes #5222

mlassnig committed:
Fix read write protocol assignment; Fix #5234

bari12 committed:
Merge pull request #5254 from joeldierkes/feature-5253-Print_the_integration_test_server_logs_in_the_actions

Testing: Print error logs in the integration test actions #5253

maany committed:
testing: enable LTS containers for release branches #5256

bari12 committed:
Database: Add query hint for 19c stabilization; Fix #5271

bari12 committed:
Merge pull request #5045 from robbarnsley/feature-4949-Added_client_REST_api_route_for_listing_and_getting_historical_requests

enhancement: added route for listing requests history #4949

bari12 committed:
Merge pull request #5237 from mlassnig/patch-5234-Clients__Fix_read_write_protocol_assignment

Clients: Fix read write protocol assignment for upload; Fix #5234

dchristidis committed:
WebUI: Fix syntax error #5278

Commit:
Transfers: store transfer paths in a new database table. Closes #5177

Until now, a multihop transfer didn't have a real database state.
Each hop of the multihop was a separate transfer request with almost
no relationship between individual requests (only a loosely coupled
json attribute). This behavior is source of multiple issues:
 - we have recurrent problems with intermediate hops living their
own life (their lifetime should be linked to the initial request
and they shouldn't be retried on their own).
 - multiple multi-hops sharing a common sub-path are not possible
and result in permanent failures of one of the transfers until the
second finishes.
 - it doesn't allow implementation of cross-tranfertool multihops

For these particular reasons, we introduce a table storing the
relationship between independent requests of a path. I change the
workflows to rely on joins with this table:
- (1) submitter only retrieves from database requests which don't have
any pending previous hops.
- (2) submitter is capable to detect that a sub-path of the current
multi-hop transfer already exists in the database. It only creates
the link with the existing sub-path and aborts submission. Submission
automatically resumes once the common sub-path is transferred thanks to
behavior mentioned in (1)
- (3) when an intermediate hop fails, the poller/receiver will also
fail the next requests in all affected paths.

bari12 committed:
Merge pull request #5068 from cserf/patch-5033-Core___provide_the_possibility_to_specify_PFNs_or_DIDs_in_declare_bad_file_replicas

Core : provide the possibility to specify PFNs or DIDs in declare_bad_file_replicas : Close #5033

bari12 committed:
Merge pull request #5280 from bari12/patch-5271-Database__Add_query_hints_for_19c

Database: Add query hint for 19c stabilization; Fix #5271

bari12 committed:
Merge pull request #5279 from dchristidis/patch-5278-WebUI__Fix_syntax_error

WebUI: Fix syntax error #5278

bari12 committed:
Merge pull request #5238 from joeldierkes/feature-5209-Add_a_Rest_Api_Documentation_linter

Documentation: Introduce the OpenApi spec generator #5209

bari12 committed:
Merge pull request #5163 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view

Feature 4992 fix broken client and rest api docs view: rule

bari12 committed:
Merge pull request #5165 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_accountlimit

REST & API: Change the `accountlimit` api docs to the OpenApi format …

bari12 committed:
Merge pull request #5173 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_accounts

REST & API: Change the `accounts` api docs to the OpenApi format #4992

bari12 committed:
Merge pull request #5201 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_credentials

REST & API: Change the `credentials` api docs to the OpenApi format #…

bari12 committed:
Merge pull request #5205 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_export

 REST & API: Change the `export` api docs to the OpenApi format #4992

bari12 committed:
Merge pull request #5236 from jamesp-epcc/patch-5235-fix-multivo-objectstore-uploads

Bug: Fix objectstore uploads on multi VO #5235

Commit:
Deletion: rewrite setting BEING_DELETED without or-chunk

joeldierkes committed:
Fix the rest api doc generation

The rest api docs use the OpenApi specification. It is strict in it's input and
requests an array with objects with the field `name` as input for parameters.

bari12 committed:
Merge pull request #5283 from joeldierkes/fix-rest-api-doc

Fix the rest api doc generation

cserf committed:
Minos/Bad replicas database inefficiency : Closes #5266

Fix tests

Fix flake8

Fix multiVO test

Code cleanup

Commit:
Transfers: support RSEs restricted to admin accounts. Closes #3510

Commit:
Deletion: duplicate delete_replicas for migration to temp tables

Commit:
Deletion: temp tables to update is_archive on collections. Closes #5166

Commit:
Deletion: temp tables for parent did iteration

Commit:
Deletion: commit in dataset test

delete_replicas tries to create a temp table each time it is called.
On oracle this is problematic if we call twice delete_replicas
within the same transaction.

joeldierkes committed:
Core: Fix emtpy array access

bari12 committed:
Merge pull request #5288 from joeldierkes/fix-master

Core: Fix emtpy array access

Commit:
Deletion: temp tables to remove did associations

Commit:
Deletion: move did_meta removal condition for oracle < 12

Commit:
Deletion: temp tables to remove dids

Commit:
Deletion: temp table to remove sources and replicas

Commit:
Deletion: temp tables to cleanup collection replicas

Commit:
Deletion: update is_archive without fetching to python

Commit:
Deletion: temp tables to remove constituents

bari12 committed:
Merge pull request #5276 from maany/patch-5256-LTS_containers_integration_tests

Select containers based on release brach for Integration Tests (master)

bari12 committed:
Merge pull request #5267 from maany/feature-5625-client_state_values

Traces: bulk updates for schemas #5265

Commit:
Deletion: de-indent code block

Commit:
Deletion: temp tables to schedule collection update

Commit:
Deletion: remove oracle hits in delete_replicas with temp tables

I dare hope that the optimizer will not need those hints now
that all queries were re-written with simple joins.

bari12 committed:
Merge pull request #5289 from cserf/patch-5266-Minos_Bad_replicas_database_inefficiency

Minos/Bad replicas database inefficiency : Closes #5266

bari12 committed:
Release: Adding newlines to release note generator

bari12 committed:
Merge pull request #5291 from bari12/patch-0-Release__Update_release_script

Release: Adding newlines to release note generator

bari12 committed:
Merge pull request #5206 from rcarpa/patch-4897-rework_list_and_mark_unlocked_replicas

Deletion: rework list_and_mark_unlocked_replicas. Closes #4897

mlassnig committed:
Remove gfalv2 implementation; Fix #5227 (#5228)

* Remove gfalv2 implementation; Fix #5227

* remove last remnants of v2

* header update

Commit:
Transfers: rework heartbeat handling. Closes #5252

Right now heartbeats are updated in the database at each iteration
(each activity for multi-activity daemons). This forces us to use
a big "older_than" value to avoid the race condition when a big bulk
size and slow access to transfertool makes hearbeats expire.
At the same time, this puts strain on database with un-needed heartbeat
updates when the daemons are lightly loaded.

Rework this behavior to perform hearbeat_handler.live() more frequently
(each submission, for example), but modify live() to only perform a
database update if enough time has passed from last update.

joeldierkes committed:
REST & API: Change the `meta` api docs to the OpenApi format rucio#4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

nimishbongale committed:
Removed hardcoded references to rucio-ui-dev.cern.ch

bari12 committed:
Merge pull request #5200 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_dirac

REST & API: Change the `dirac` api docs to the OpenApi format #4992

bari12 committed:
Merge pull request #5294 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_config

 REST & API: Change the `config` api docs to the OpenApi format #4992

bari12 committed:
Merge pull request #5295 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_heartbeats

 REST & API: Change the `heartbeats` api docs to the OpenApi format #…

bari12 committed:
Merge pull request #5296 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_identities

 REST & API: Change the `identity` api docs to the OpenApi format #4992

bari12 committed:
Merge pull request #5299 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_meta

REST & API: Change the `meta` api docs to the OpenApi format rucio#4992

maany committed:
webui: add link to survey #5305

cfenell committed:
WebDAV bugfixes

cfenell committed:
Reverted debug print

cfenell committed:
Reverted debug print

mlassnig reviewed (commented):
So the code looks good, however I don't understand why the automated test suite did not run. @bari12 ?

joeldierkes commented:
"First-time contributors need a maintainer to approve running workflows."
Don't you have the 'Approve and run' button?

mlassnig commented:
> "First-time contributors need a maintainer to approve running workflows." Don't you have the 'Approve and run' button?

nope /shrug :-D

bari12 commented:
The approve workflow button should be there for anyone in @rucio/rucio-dev 🤷‍♂️ 😄

cserf committed:
Client : Upload broken for cloud resources : Closes #5311

Commit:
Core & Internals: use third_party_copy_read/write. Closes #3254

There is a desire to have different configurations for an RSE
when used as source or destination in a transfer. The old
third_party_copy doesn't allow that, so the _read/_write was
introduced to the database at some moment, but were ignored
by rucio code.

Switch to using these setting by default. Add a database migration
to copy the values from the existing field. Also ensure that
setting third_party_copy only when creating a protocol results in
the population of all 3 fields.

joeldierkes committed:
REST & API: Change the `import` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

joeldierkes committed:
REST & API: Change the `lifetime exception` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

cfenell committed:
Documentation: server install from source

cfenell committed:
removed LaTeX formatting from md

Commit:
Core & Internals: reduce default cache expiration lifetime. Closes #5303

Reduce to 15m where it was >15m.

Also fix a bug where it wasn't correctly passed through and a default
value of 3600 was used for creating memcached region.

cfenell committed:
Update rucio-server.md

joeldierkes committed:
REST & API: Change the `locks` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

bari12 committed:
Merge pull request #5312 from cserf/patch-5311-Client___Upload_broken_for_cloud_resources

Client : Upload broken for cloud resources : Closes #5311

bari12 committed:
Merge pull request #5301 from nimishbongale/patch-5171-remove-hardcoded-reference-to-rucio-ui-dev.cern.ch

Removed hardcoded references to rucio-ui-dev.cern.ch

bari12 committed:
Merge pull request #5314 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_lifetime_exceptions

REST & API: Change the `lifetime exception` api docs to the OpenApi f…

bari12 committed:
Merge pull request #5316 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_locks

REST & API: Change the `locks` api docs to the OpenApi format #4992

bari12 committed:
Merge pull request #5318 from rcarpa/patch-5303-reduce_cache_duration

Core & Internals: reduce default cache expiration lifetime. Closes #5303

bari12 committed:
Merge pull request #5307 from maany/feature-5305-add_webui_survey_link

WebUI: add link to survey

cfenell committed:
Update rucio-server.md

bari12 committed:
Release: Requirements change 1.28

cserf committed:
Client : Add add-exception command in the CLI : Closes #5292

Commit:
Replicas: don't try to download geoip db without licence. Closes #5233

bari12 committed:
Release: Fix release note script

Commit:
Replicas: factorize almost identical metalink generation code. #4981

Also fixed a bug where the limit wasn't correctly enforced on json
outputs for get requests.

Commit:
Replicas: separate api replica sorting and response generation. #4981

bari12 committed:
Merge pull request #5323 from bari12/patch-0-Release__Fix_release_note_script

Release: Fix release note script

bari12 committed:
Merge pull request #5321 from rcarpa/patch-5233-skip_geoip_download_if_no_licence

Replicas: don't try to download geoip db without licence. Closes #5233

bari12 committed:
Merge pull request #5309 from rcarpa/patch-5297-memcached_for_auth_tokens

Core & Internals: allow external cache for auth tokens. Closes #5297

bari12 committed:
Merge pull request #5304 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_archives

 REST & API: Change the `archive` api docs to the OpenApi format #4992

maany committed:
testing: py3 fts-rest-client #5257

maany committed:
testing: handle edge cases for integration test container selection #5256

cfenell committed:
Merge branch 'master' of github.com:rucio/rucio

joeldierkes committed:
REST & API: Change the `ping` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

bari12 committed:
Merge pull request #5269 from rcarpa/patch-4873-remove_fts_specifics_poller

Transfers: disentangle fts3 code from poller/receiver. Closes #4873. Closes #4576

joeldierkes committed:
REST & API: Change the `nongrid traces` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

Commit:
REST & API: decorate api/rule with session managers. #5002

Only create the session object once per api call to rules enpoint
and pass it through. The goal is to test if the number of un-needed
rollbacks on the database side is decreased thanks to avoiding
sessionmaker.remove() calls.

Commit:
REST & API: decorate api/replica with session managers. #5002

Commit:
REST & API: decorate api/did with session managers. #5002

joeldierkes committed:
REST & API: Change the `redirect` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

cserf committed:
Dirac : Provide the possibility to set a lifetime on the default rule : Closes #5164

maany committed:
Merge branch 'master' into feature-5257-py3_fts_rest_client

cfenell committed:
verify always (<ca path>|True|False)

cfenell committed:
Merge branch 'master' of github.com:rucio/rucio

cfenell committed:
Merge branch 'master' of github.com:rucio/rucio

cfenell committed:
Merge branch 'master' of github.com:cfenell/rucio

bari12 committed:
Merge pull request #5293 from rcarpa/patch-4793-temp_tables_to_delete_replicas

 Temp tables to delete replicas. #4793. Closes #5166

joeldierkes committed:
Always print the integration test logs #5253

If any job before the `Print server error logs` fails, the job itself will not
run. Since these are the cases in which the logs can provide additional
information, they should be printed.

bari12 committed:
Merge pull request #5328 from maany/feature-5257-py3_fts_rest_client

Testing: add py3 fts rest client + integration test improvements

bari12 committed:
Merge pull request #5337 from joeldierkes/feature-5253-Print_the_integration_test_server_logs_in_the_actions_2

Always print the integration test logs #5253

bari12 committed:
Merge pull request #5282 from rcarpa/patch-5177-transfers_hop_table

Transfers: store transfer paths in a new database table. Closes #5177

bari12 committed:
Merge pull request #5298 from rcarpa/patch-3510-prevent_getting_files_from_gcs

Transfers: support RSEs restricted to admin accounts. Closes #3510

piperov committed:
Consistency checks: add storage-consistency-actions #5113 (#5120)

* Consistency checks: add storage-consistency-actions #5113

* add daemon_sleep

* switch to using record_counter

* add prometheus counters

* attempt at pylint

* implemented the requested changes from the pull-request

* respond to second and third waves of reviews

* a round of beautification

* rebase to master

* beautification

* more beautification

* addressing latest round of reviwers comments

* Change interface __exists_replicas to __exist_replicas

* undefined path, whitespaces

* missing whitespace around operator

* removing explicit cms mentions

* removing explicit cms mention

* remove get_rse_vo

* fix header

robbarnsley committed:
redirect_uri now takes into account paths #5220 (#5325)

* redirect_uri now takes into account paths

* removed header duplication

bari12 commented on doc/rucio-server.md:
```diff
@@ -0,0 +1,590 @@
+Setting up a Rucio server
```
Thanks, this is very nice! 😄 
I think it should go to https://github.com/rucio/documentation though, since all the documentation goes there.
We need to find a good point to embed it though, since this is just one way how to do it. (The container one being a different one)

bari12 commented on doc/rucio-server.md:
```diff
@@ -0,0 +1,590 @@
+
+* **itsdangerous** is too new in PyPI but a compatible package exists in Ubuntu 20.04 
+     
+    $sudo apt install python3-itsdangerous
```
I think you can remove that one. From 1.28 onwards it's not needed anymore. This was just due to an incompatibility between the flask version used and itsdangerous.

bari12 reviewed (commented):
@mlassnig @cserf please review the webdav part of this.
@cfenell for the header-lint failing, you either need to update the headers of the modules manually (and correctly) or call the `tools/add_header -i` script on them, to update them manually (Add you as an author etc.)

bari12 committed:
Merge pull request #5326 from rcarpa/patch-4981-refactor_replica_api_sorting

Replicas: refactor api sorting and response generation. Closes #4981

bari12 committed:
Merge pull request #5331 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_ping

REST & API: Change the `ping` api docs to the OpenApi format #4992

bari12 committed:
Merge pull request #5332 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_redirect

REST & API: Change the `redirect` api docs to the OpenApi format #4992

bari12 committed:
Merge pull request #5333 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_nongrid_traces

REST & API: Change the `nongrid traces` api docs to the OpenApi forma…

Commit:
Replicas: remove whitespaces from geoip cache key

Memcached doesn't work with spaces in keys.

bari12 committed:
Merge pull request #5336 from cserf/patch-5164-Dirac___Provide_the_possibility_to_set_a_lifetime_on_the_default_rule

Dirac : Provide the possibility to set a lifetime on the default rule…

bari12 committed:
Merge pull request #5118 from jamesp-epcc/feature-4918-client-policy-packages

Clients: allow policy packages on client #4918

cserf commented on lib/rucio/rse/protocols/webdav.py:
```diff
@@ -51,7 +54,11 @@ def init_poolmanager(self, connections, maxsize, block=False):
                                        block=block,
-                                       ssl_version=ssl.PROTOCOL_TLSv1)
+                                       cert_reqs="CERT_REQUIRED",
+                                       ca_cert_dir="/etc/grid-security/certificates",
```
What about checking `X509_CERT_DIR` ?

cserf reviewed (commented):
Looks good to me except the flake8 errors.
I also have a comment on the use of X509_CERT_DIR instead of hardcoding the cert path

Commit:
Monitoring & Logging: upgrade prometheus_client. 5345

https://github.com/prometheus/client_python/releases
- remove labelvalues, which we didn't use anyway
- support upper-case multiproc variable name.
Lower-case is now deprecated

bari12 committed:
Merge pull request #5338 from rcarpa/patch-0-remove_replica_sorter_cache_spaces

Replicas: remove whitespaces from geoip cache key

bari12 committed:
Merge pull request #5315 from rcarpa/patch-3254-third_party_copy_read_write

Core & Internals: use third_party_copy_read/write. Closes #3254

Commit:
Deletion: use CASE in temp table list_replicas

Otherwise the oracle optimiser doesn't use the functional index.

It's actually how the query was written initially,
https://github.com/rucio/rucio/blob/9a59596ae365fbd79c51d0f36c851cc3265634a8/lib/rucio/core/replica.py#L2756
but I didn't know about the reason for it, so I introduced a
regression on the temporary table query.

mlassnig committed:
Update docker versions; Fix #4198 (#5324)

* Update dev docker databases; Fix #4198

* finalised configuration

* header fix

faluchet committed:
Consistency checks: Drop unmaintened bz2file module, use plain bz2 instead - Fix #5348

robbarnsley committed:
added docker-compose-storage-externalmetadata compose file with mongodb service
integration tests now use docker-compose-storage-externalmetadata compose file
added pymongo client to requirements

cfenell committed:
unnecessary ifs before pfn2path removed

cfenell committed:
Merge branch 'master' of https://www.github.com/rucio/rucio

cfenell committed:
Merge branch 'master' of github.com:cfenell/rucio

cfenell committed:
removed debug print

maany committed:
Add nightly build, remove py37,38  (#5350)

* testing: add nightly build, remove py37,38 #5349

* testing: add py3738 test suite

bari12 committed:
Merge pull request #4736 from cserf/patch-4706-Rebalancing___BB8_rewrite

Rebalancing : BB8 rewrite : Closes #4706

bari12 committed:
Merge pull request #5351 from rcarpa/patch-0-fix_temp_table_list_delete_replicas_query

Deletion: use CASE in temp table list_replicas

bari12 committed:
Merge pull request #5339 from bari12/feature-0-Release__Dependency_upgrade_for_1_28

Release: Requirements change 1.28

bari12 committed:
Merge pull request #5003 from rcarpa/patch-5002-decorate_with_session_api_rules

REST & API: decorate some APIs with session managers. Closes #5002

bari12 committed:
Merge pull request #5353 from rcarpa/patch-5345-prometheus_client_upgrade

Monitoring & Logging: upgrade prometheus_client. 5345

bari12 committed:
Core: Add timeout for vp query; Fix #5360

All other VP queries have a timeout to protect from the service being
unavailable. This query did not.

bari12 committed:
Merge pull request #5361 from bari12/patch-5360-Core__Enforce_timeout_for_VP

Core: Add timeout for vp query; Fix #5360

bari12 committed:
Merge pull request #5306 from rcarpa/patch-5252-conveyor_heartbeat_handling

Transfers: rework heartbeat handling. Closes #5252

bari12 committed:
Merge pull request #5322 from cserf/patch-5292-Client___Add_add_exception_command_in_the_CLI

Client : Add add-exception command in the CLI : Closes #5292

bari12 commented:
HI @cfenell 
Still a few flake8 errors, you can have a look at them directly, if you check the test jobs. (Checks)

cfenell committed:
Merge branch 'master' of github.com:cfenell/rucio

cfenell committed:
ignore generated pdf

cfenell committed:
gfal2 documentation

cfenell committed:
Merge branch 'master' of https://www.github.com/rucio/rucio

cfenell committed:
gfal2 documentation typos

cfenell committed:
gfal doc: note about host certs

cfenell committed:
gfal doc: add to note about host certs

cfenell committed:
gfal doc format

cfenell committed:
gfal doc format

bari12 committed:
Merge pull request #5017 from joeldierkes/feature-4995-Allow_overriding_all_possible_options_in_move_rule

Core: allow `move_rule` to override all options Fix #4995

bari12 committed:
Merge pull request #5261 from robbarnsley/patch-5260-add_pymongo_to_requirements_and_mongodb_container_to_CI_pipeline

enhancement: requirements for mongodb to be included as an external metadata provider in CI pipeline #5260

bari12 committed:
Merge pull request #5352 from faluchet/patch-5348-drop_bz2file

Consistency checks: Drop unmaintened bz2file module, use plain bz2 in…

cfenell committed:
Merge branch 'master' of https://www.github.com/rucio/rucio

cfenell committed:
Merge branch 'master' of github.com:cfenell/rucio

maany committed:
Testing: Add VO Test framework (#5286)

* policy package build matrix

* Testing: add cli for processing policy package build matrix #3878

* Testing: add verbose mode, installation_cmd #3878

* Testing:  build_images.py compatibility votest #3878

* Testing: fix build_images.py #3878

* Testing: update python datatype in matrix #3878

* Testing: update policy package matrix #3878

* testing: update rucio config in votest_helper #3878

* testing: override policy config in rucio.cfg #3878

* testing: add test filter #3878

* testing: collect tests in votest_helper #3878

* testing: filter and validate tests to run #3878

* testing: configure run_test_docker and test.sh for votest #3878

* testing: add votest workflow #3878

* testing: disable autotests, integration tests #3878

* testing: run tests, fix headers #3878

* testing: rename votest workflow, atlas tests #3878

* testing: fix default tests #3878

* testing: introduce composite actions #3878

* testing: add atlas lfn2prn algorithm #3878

* testing: fix test_rse #3878

* testing: exclude flaky atlas tests, add headers, lint #3878

* testing: add postgres14, minor fixes #3878

cfenell committed:
add_header and pyflakes checks

cfenell commented:
Ran tools/add_header
Deleted unused "import ssl"

bari12 committed:
Merge pull request #5313 from joeldierkes/feature-4992-Fix_broken_Client_and_Rest_API_docs_view_import

REST & API: Change the `import` api docs to the OpenApi format #4992

cfenell committed:
Fixed outstanding PEP8 issues and typos

cfenell committed:
Reverted header

cfenell commented:
There were still a few PEP errors in the code, spotted by flake8 but not pyflakes3- I mixed up those two programs.

cfenell committed:
Merge branch 'master' of github.com:rucio/rucio

cfenell committed:
Headers fixed

cfenell committed:
typo fixes in verify check

cfenell committed:
Fixed headers
Reset author

cfenell committed:
WIP header fixes

cfenell committed:
Headers and git history fixed to match

cfenell committed:
Merge branch 'master' of github.com:rucio/rucio
